datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restoreindices/SendMAEStep.java

```
@@ -131,18 +133,45 @@ public Function<UpgradeContext, UpgradeStepResult> executable() {
 
       List<Future<RestoreIndicesResult>> futures = new ArrayList<>();
       startTime = System.currentTimeMillis();
-      while (start < rowCount) {
-        args = args.clone();
-        args.start = start;
-        futures.add(executor.submit(new KafkaJob(context, args)));
-        start = start + args.batchSize;
-      }
-      while (futures.size() > 0) {
-        List<RestoreIndicesResult> tmpResults = iterateFutures(futures);
-        for (RestoreIndicesResult tmpResult: tmpResults) {
-          reportStats(context, finalJobResult, tmpResult, rowCount, startTime);
+      if (args.urnBasedPagination) {
+        RestoreIndicesResult previousResult = null;
+        int rowsProcessed = 1;
+        while (rowsProcessed > 0) {
+          args = args.clone();
+          if (previousResult != null) {
+            args.lastUrn = previousResult.lastUrn;
+            args.lastAspect = previousResult.lastAspect;
+          }
+          args.start = start;
+          context.report().addLine(String.format("Getting next batch of urns + aspects, starting with %s - %s", args.lastUrn, args.lastAspect));
+          Future<RestoreIndicesResult> future = executor.submit(new KafkaJob(context, args));
+          while (!future.isDone()) { }
+          try {
+            RestoreIndicesResult result = future.get();
+            reportStats(context, finalJobResult, result, rowCount, startTime);
+            previousResult = result;
+            rowsProcessed = result.rowsMigrated + result.ignored;
+            context.report().addLine(String.format("Rows processed this loop %d", rowsProcessed));
+            start += args.batchSize;
+          } catch (InterruptedException | ExecutionException e) {
+            e.printStackTrace();
```

This should be added to the context report and treated as a failure condition rather than printing to std out.